### DISCUSSION TITLE
Dynamic joins from string definition

### DISCUSSION THREAD

**Discussion by mmeisso** [sqlalchemy/sqlalchemy#10887] [Usage Questions]
Hi, I try to convert "dotpath" to a chain of joins. While I have some results, I still have some issues.
In short, I need to apply this dotpath twice on a query, one on the base model ( the FROM ), an other time on an alias of the base Model, like :
```sql
SELECT * FROM model
INNER JOIN model AS anchor on anchor.id = model.id

JOIN model_associations AS model_associations_1 ON anchor
JOIN model AS model_1 ON model_associations_1

JOIN model_associations model_associations_2 ON model
JOIN model model_2 ON model_associations_2

# Then later, something like WHERE model > anchor
```

```python
def path_to_models(field: string, base_model: Model):
    dotted_field = field.split(".")
    previous_model = base_model
    compounds = []

    for df in dotted_field[:-1]: # Skip the field for now
        target_model = getattr(previous_model, df, None)
        if target_model is None:
            break
        compounds.append(target_model)

def join_from(dot_path: string, query: Query, base_model):
    item_fields = path_to_models(
        dot_path, base_model
    )

    for compound in item_fields:
        query = query.join(compound)

    alias = aliased(base_model, "paginate_after")
    query = self.join(alias, alias.id == last_item_id)
    alias_fields = path_to_models(
        dot_path, alias
    )
    for compound in alias_fields:
        query = query.join(compound)
    raise Exception(query) # Poor man debug tool

query = session.query(Model)

this_id = 1
join_from("model_associations.model.name", query, Model)

```

But I don't have the join on base model done it all atm... I tried various versions, it results in only joining from alias

**Comment by CaselIT** [comment]
Hi,

I don't think you meant to do this: `query = self.join(alias, alias.id == last_item_id)` since you are replacing query without taking into consideration the previous joins.
Other than that from a quick look the code seems mostly ok

**Comment by mmeisso** [comment]
You are right ans I got it fixed, but I still miss joins :/
________________________________

**Comment by CaselIT** [comment]
Looking better at your function `path_to_models` is not correct since a you can't access model fields from a relationship. 

You can get the relationsiphip property using `.property`, from it you can get the mapper and from it the related class.

Something like this :
```py
def path_to_models(field, base_model):
    dotted_field = field.split(".")
    previous_model = base_model
    compounds = []

    for df in dotted_field[:-1]: # Skip the field for now
        target_attr = getattr(previous_model, df, None)
        if target_attr is None:
            break
        compounds.append(target_attr)
        previous_model = target_attr.property.parent.class_
    return compounds
```

docs:
- https://docs.sqlalchemy.org/en/20/orm/internals.html#sqlalchemy.orm.InstrumentedAttribute
- https://docs.sqlalchemy.org/en/20/orm/internals.html#sqlalchemy.orm.MapperProperty.parent
- https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.Mapper
